src/Filesystem/Filesystem.php

```diff
@@ -33,6 +35,42 @@ public function add(string $file, Media $media, ?string $targetFileName = null)
         app(FileManipulator::class)->createDerivedFiles($media);
     }
 
+    public function addRemote(RemoteFile $file, Media $media, ?string $targetFileName = null)
+    {
+        $this->copyToMediaLibraryFromRemote($file, $media, null, $targetFileName);
+
+        event(new MediaHasBeenAdded($media));
+
+        app(FileManipulator::class)->createDerivedFiles($media);
+    }
+
+    public function copyToMediaLibraryFromRemote(RemoteFile $file, Media $media, ?string $type = null, ?string $targetFileName = null)
+    {
+        $storage = Storage::disk($file->getDisk());
+
+        $destinationFileName = $targetFileName ?: $file->getFilename();
+
+        $destination = $this->getMediaDirectory($media, $type).$destinationFileName;
+
+        if ($file->getDisk() === $media->getDiskDriverName()) {
+            $storage->copy($file->getKey(), $destination);
+        } else {
+            $contents = $storage->get($file->getKey());
```

Comment:
Wouldn't this be problematic when using very large files? Shouldn't we use streams here?

Reply:
Updated